### ISSUE TITLE
Example involving parameter bounds and NormalisingFlow

### ISSUE THREAD

**Issue by itsdfish** [msainsburydale/NeuralEstimators.jl#38]
Hello!

I am excited to learn more about your package. Thank you very much for making it available. 

I am interested in applying it in my line of work, so I am trying to scale up the first example in the documentation to something more difficult. Right now my first step is to use `NormalizingFlow` to estimate the parameters of the simple Gaussian model in the docs. I am encountering a dimension mismatch because I am not sure how to configure the neural network (I have a basic understanding of MLPs and CNNs, but not more complex networks). I was wondering if you could provide some guidance on the code below? If you think it would be useful to include as an example in the docs, I can submit a PR. 

Thanks!

```julia
using Distributions
using Flux
using NeuralEstimators


n = 100    # dimension of each data replicate (univariate)
d = 2        # dimension of the parameter vector θ
w = 128   # width of each hidden layer 

function sample(K)
	μ = rand(Normal(0, 10), K)
	σ = rand(Uniform(0, 10), K)
	θ = vcat(μ', σ')
	return θ
end

simulate(θ, m = 100) = [ϑ[1] .+ ϑ[2] * randn(n, m) for ϑ ∈ eachcol(θ)]  

approx_dist = NormalisingFlow(d, 2 * d)
final_layer = Parallel(
    vcat,
    Dense(w, 1, identity),     # μ ∈ ℝ
    Dense(w, 1, softplus)      # σ > 0
)

# Inner and outer networks
ψ = Chain(Dense(n, w, relu), Dense(w, d, relu))    
ϕ = Chain(Dense(d, w, relu), final_layer)           

network = DeepSet(ψ, ϕ)

# Initialise a neural point estimator
estimator = PosteriorEstimator(approx_dist, network) 

# Train the estimator
estimator = train(estimator, sample, simulate, epochs = 20; m)

# Assess the estimator
θ_test = sample(1000)
Z_test = simulate(θ_test)
assessment = assess(estimator, θ_test, Z_test)
bias(assessment)   # μ = 0.001, σ = 0.001
rmse(assessment)   # μ = 0.05,  σ = 0.04

# Apply the estimator to observed data
θ = [0.0, 1.0]          # true parameters
Z = simulate(θ, m)      # "observed" data
post_samples = sampleposterior(estimator, Z)
````

**Comment by msainsburydale** [comment]
Hi! Thanks for your interest in the package and your question. 

When using a normalising flow, the neural network needs to map data to summary statistics of the same dimension specified when initialising the `NormalisingFlow` object (`2d` in your code). However, the network above using `final_layer` outputs a summary statistic of dimension `d` rather than `2d`, hence the dimension mismatch. 

The following code should work. 

```julia
using Distributions
using Flux

# Update to current version of NeuralEstimators:
using Pkg; Pkg.add(url = "https://github.com/msainsburydale/NeuralEstimators.jl")
using NeuralEstimators

n = 100   # dimension of each data replicate 
m = 100   # number of independent replicates 
d = 2     # dimension of the parameter vector θ
w = 128   # width of each hidden layer 

function sample(K)
	μ = rand(Normal(0, 10), K)
	σ = rand(Uniform(0, 10), K)
	θ = vcat(μ', σ')
	return θ
end

simulate(θ, m) = [ϑ[1] .+ ϑ[2] * randn(n, m) for ϑ ∈ eachcol(θ)]  

# Approximate distribution
approx_dist = NormalisingFlow(d, 2d)

# Neural network mapping data to summary statistics (of the same dimension used in the approximate distribution)
ψ = Chain(Dense(n, w, relu), Dense(w, w, relu))    
ϕ = Chain(Dense(w, w, relu), Dense(w, 2d))           
network = DeepSet(ψ, ϕ)

# Initialise a neural posterior estimator
estimator = PosteriorEstimator(approx_dist, network) 

# Train the estimator
estimator = train(estimator, sample, simulate, epochs = 20, m = m)

# Assess the estimator
θ_test = sample(1000)
Z_test = simulate(θ_test)
assessment = assess(estimator, θ_test, Z_test)
bias(assessment)   # μ = 0.001, σ = 0.001
rmse(assessment)   # μ = 0.05,  σ = 0.04

# Apply the estimator to observed data
θ = [0.9, 2.0]       # true parameters
Z = simulate(θ, m)      # "observed" data
post_samples = sampleposterior(estimator, Z)
```

Note that in this specific example it would make more sense to set `n=1` and `m=100*100`, but I assume you plan to try it next with a multivariate model (`n > 1`).

**Comment by itsdfish** [comment]
Thank you for your help. I was uncertain whether 
```julia 
ϕ = Chain(Dense(w, w, relu), Dense(w, 2d))
 ```
would respect the bounds for $\sigma$. The modifications you made worked, and mostly kept the posterior of $\sigma$ greater than zero ($\approx 99.3\%$ of cases). 

I am trying to adapt your code above to a simple decision making model called the Log Normal Race (LNR). In a simple version, each observation is a choice $$c \in \\{1,2 \\}$$ for option 1 and option 2, and a reaction time $t > 0$. It has the following parameters: a $2 \\times1$ vector $\mu$, a scalar $\sigma >0$, and an additive constant $\tau >0$. 

$(c,t) \sim \mathrm{LNR}(\boldsymbol{\mu}, \sigma, \tau)$

The LNR is similar to the Levy-Flight model in the paper [BayesFlows](https://arxiv.org/pdf/2003.06281), but likely easier to estimate.  The code below runs without errors, but the posterior distributions often do not make sense. Sometimes they are overly wide, other times, they are not very sensitive to changes in the data `Z = simulate(θ, m)`. The assessment plots frequently display outliers 7-12 orders of magnitude greater than what would be expected from the prior distributions. The outliers sometimes occur for high density values of the prior.  The network and training parameters seem consistent with the paper above. Changing the number of training sets $K$, the number of epochs (which usually terminates early), the $w$ parameter, and d* in  `NormalisingFlow` does seem to help much. I was wondering if you might have any idea what I am doing incorrectly. 

Thanks!


```julia
using AlgebraOfGraphics
using CairoMakie
using Distributions
using Flux
using NeuralEstimators
using Plots
using SequentialSamplingModels

n = 2           # dimension of each data replicate 
m = 100         # number of independent replicates 
d = 4           # dimension of the parameter vector θ
w = 128         # width of each hidden layer 

function sample(K)
	ν = rand(Normal(-2, 2), K, 2)
	σ = rand(truncated(Normal(1, 3),0, Inf), K)
    τ = rand(Uniform(.200, .300), K)
	θ = vcat(ν', σ', τ')
	return θ
end

to_array(x) = [x.choice'; x.rt']
simulate(θ, m) = [to_array(rand(LNR(ϑ[1:2], ϑ[3], ϑ[4]), m)) for ϑ ∈ eachcol(θ)]  

# Approximate distribution
approx_dist = NormalisingFlow(d, 2d)

# Neural network mapping data to summary statistics (of the same dimension used in the approximate distribution)
ψ = Chain(Dense(n, w, relu), Dense(w, w, relu))    
ϕ = Chain(Dense(w, w, relu), Dense(w, 2d))           
network = DeepSet(ψ, ϕ)

# Initialise a neural posterior estimator
estimator = PosteriorEstimator(approx_dist, network) 

# Train the estimator
estimator = train(estimator, sample, simulate, epochs = 20, m = m, K = 20_000)

# Assess the estimator
θ_test = sample(1000)
Z_test = simulate(θ_test, m)
assessment = assess(estimator, θ_test, Z_test)
bias(assessment)  
rmse(assessment) 
AlgebraOfGraphics.plot(assessment)


# Apply the estimator to observed data
θ = [-1,-2,.5,.250]           # true parameters
Z = simulate(θ, m)             # "observed" data
post_samples = sampleposterior(estimator, Z)
Plots.histogram(post_samples', layout = (4,1))
vline!([θ'])
```

**Comment by msainsburydale** [comment]
You are right that 

```julia
ϕ = Chain(Dense(w, w, relu), Dense(w, 2d))
```

will not necessarily respect the bounds of the parameter space. I suppose that you can just discard samples that are not valid. However I'll also give it some thought and try to add some functionality that will ensure that all samples are valid. 

I believe that the issue is that your simulated data can be very large in absolute value. This can cause problems when training the neural networks due to numerical instability. Since your data are positive, a log transformation can be used (see the line that constructs `ψ` below). (Note also that I changed `to_array()` since the original definition wasn't running for some reason.) 

```julia
using AlgebraOfGraphics
using CairoMakie
using Distributions
using Flux
using NeuralEstimators
using Plots
using SequentialSamplingModels

n = 2           # dimension of each data replicate 
m = 100         # number of independent replicates 
d = 4           # dimension of the parameter vector θ
w = 128         # width of each hidden layer 

function sample(K)
	ν = rand(Normal(-2, 2), K, 2)
	σ = rand(truncated(Normal(1, 3),0, Inf), K)
        τ = rand(Uniform(.200, .300), K)
	θ = vcat(ν', σ', τ')
	return θ
end

to_array(x) = Float32.(hcat((collect).(x)...))
simulate(θ, m) = [to_array(rand(LNR(ϑ[1:2], ϑ[3], ϑ[4]), m)) for ϑ ∈ eachcol(θ)]  

# Approximate distribution
approx_dist = NormalisingFlow(d, 2d)

# Neural network mapping data to summary statistics (of the same dimension used in the approximate distribution)
ψ = Chain(x -> log.(x), Dense(n, w, relu), Dense(w, w, relu)) # NB now using log-transform for numerical stability
ϕ = Chain(Dense(w, w, relu), Dense(w, 2d))           
network = DeepSet(ψ, ϕ)

# Initialise a neural posterior estimator
estimator = PosteriorEstimator(approx_dist, network) 

# Train the estimator
estimator = train(estimator, sample, simulate, m = m, K = 25_000)

# Assess the estimator
θ_test = sample(1000)
Z_test = simulate(θ_test, m)
assessment = assess(estimator, θ_test, Z_test)
bias(assessment)  
rmse(assessment) 
AlgebraOfGraphics.plot(assessment)

# Apply the estimator to observed data
θ = [-1,-2,.5,.250]       # true parameters
Z = simulate(θ, m)      # "observed" data
post_samples = sampleposterior(estimator, Z)
Plots.histogram(post_samples', layout = (4,1))
vline!([θ'])
```

<img width="703" alt="Image" src="https://github.com/user-attachments/assets/461d80ee-cf10-43c3-bbed-4de03c24bbb7" />

<img width="772" alt="Image" src="https://github.com/user-attachments/assets/4ceb0ef9-51af-4b80-a4a2-881b84149cba" />

**Comment by itsdfish** [comment]
Awesome! This works great now. 

> However I'll also give it some thought and try to add some functionality that will ensure that all samples are valid.

That would be great


> (Note also that I changed `to_array()` since the original definition wasn't running for some reason.)

Interestingly, the change you made does not work for me. Maybe the package manager used an old version of `SequentialSamplingModels`.

**Comment by bgroenks96** [comment]
I think the problem of handling parameter bounds could probably be resolved by integrating with [SimulationBasedInference.jl](https://github.com/bgroenks96/SimulationBasedInference.jl).
